cdc/api/v2/safepoint.go

```
@@ -0,0 +1,225 @@
+		return nil, err
+	}
+	return h.helpers.getPDSafepoint(up.PdEndpoints)
+}
+
+// querySafePoint request and returns safepoints from PD
+// @Summary Get the safepoint information of a TiCDC node
+// @Description This API is a synchronous interface. If the request is successful,
+// the safepoint information of the corresponding pd is returned.
+//
+// @Tags common,v2
+// @Produce json
+// @Success 200 {object} SafePoint
+// @Failure 500,400 {object} model.HTTPError
+// @Router	/api/v2/safepoint [get]
+func (h *OpenAPIV2) querySafePoint(c *gin.Context) {
+	safePoint, err := h.getPDSafepoint()
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	c.IndentedJSON(http.StatusOK, safePoint)
+}
+
+// setSafePoint request and returns safepoints from PD
+// @Summary Set the safepoint information of a TiCDC node
+// @Description This API is a synchronous interface. If the request is successful,
+// the safepoint information of the corresponding pd is returned.
+//
+// @Tags common,v2
+// @Accept json
+// @Produce json
+// @Success 200 {object} SafePoint
+// @Failure 500,400 {object} model.HTTPError
+// @Router	/api/v2/safepoint [post]
+func (h *OpenAPIV2) setSafePoint(c *gin.Context) {
+	ctx := c.Request.Context()
+
+	safePointConfig := &SafePointConfig{}
+	if err := c.BindJSON(safePointConfig); err != nil {
+		_ = c.Error(cerror.WrapError(cerror.ErrAPIInvalidParam, err))
+		return
+	}
+	if safePointConfig.TTL <= 0 {
+		_ = c.Error(cerror.WrapError(cerror.ErrAPIInvalidParam, errors.New("can't set ttl<=0")))
+		return
+	}
+	serviceID := h.getServiceID(safePointConfig.ServiceIDSuffix)
+
+	resp := &SafePoint{}
+	err := h.withUpstreamConfig(ctx, &UpstreamConfig{},
+		func(ctx context.Context, client pd.Client) error {
+			minServiceSafePoint, err := client.UpdateServiceGCSafePoint(ctx, serviceID, safePointConfig.TTL, safePointConfig.StartTs)
+			if err != nil {
+				return err
+			}
+			if minServiceSafePoint > safePointConfig.StartTs {
+				// use minServiceSafePoint to set
```

I think we should return an ErrCliInvalidServiceSafePoint error here and not try to set a different safepoint silently.

---

This command is used to prevent GC. It is inconvenient for users to specify a proper safepoint.

---

User can simple use one changefeed's checkpointTs as a safepoint.

---

How about `minServiceSafePoint > safePointConfig.StartTs && safePointConfig.StartTs == 0`? If users don't specify a safepoint, the minServiceSafePoint will be used. The others will return an error.  We define [defaultServiceIDSuffix](https://github.com/pingcap/tiflow/pull/11471/files#diff-0d2d4307b29058ba0f8afd647659b99f6fa28400848ff5796851b556d238dbfeR36) to let users can't update original serviceSafePoints.